sw/device/silicon_creator/lib/base/sec_mmio.h

```diff
@@ -0,0 +1,189 @@
+   */
+  uint32_t values[kSecMmioRegFileSize];
+
+  /**
+   * List of expected register addresses.
+   */
+  uint32_t addrs[kSecMmioRegFileSize];
+
+  /**
+   * Represents the expected number of register values.
+   */
+  uint32_t last_index;
+  /**
+   * Represents the number of register write operations. Incremented by the
+   * `sec_mmio_write32()` function.
+   */
+  uint32_t write_count;
+  /**
+   * Represents the expected number of register write operations. Incremented by
+   * the `sec_mmio_write_increment()` function.
+   */
+  uint32_t expected_write_count;
+  /**
+   * Represents the number of times the check functions have been called.
+   * Incremented by the `sec_mmio_check_values()` and the
+   * `sec_mmio_check_counters()` functions.
+   */
+  uint32_t check_count;
+} sec_mmio_ctx_t;
+
+/**
+ * Shutdown module callback handler.
+ */
+typedef void (*sec_mmio_shutdown_handler)(void);
+
+/**
+ * Initializes the module.
+ *
+ * Registers the `callee` callback handler and initializes the internal
+ * `sec_mmio_ctx_t` context.
+ */
+void sec_mmio_init(sec_mmio_shutdown_handler callee);
+
+/**
+ * Reads an aligned uint32_t from the MMIO region `addr`.
+ *
+ * This function implements a read-read-comparison operation. The first read
+ * is stored in the list of expected register values for later comparison
+ * via `sec_mmio_check_values()`.
+ *
+ * A shutdown sequence is initiated if the comparison operation fails.
+ *
+ * @param addr The address to read from.
+ * @return the read value.
+ */
+uint32_t sec_mmio_read32(uint32_t addr);
+
+/**
+ * Writes an aligned uint32_t to the MMIO region `base` at the give byte
+ * `offset`.
+ *
+ * This function implements a write-read-comparison operation. The first write
+ * value is stored in the list of expected register values for later comparison
+ * via `sec_mmio_check_values()`.
+ *
+ * On successful calls, this function will increment the internal count of
+ * writes. The caller is responsible to setting the expected write count by
+ * calling `sec_mmio_write_increment()`.
+ *
+ * A shutdown sequence is initiated if the comparison operation fails.
+ *
+ * @param addr The address to write to.
+ * @param value The value to write.
+ */
+void sec_mmio_write32(uint32_t addr, uint32_t value);
+
+/**
+ * Increment the expected count of register writes by `value`.
+ *
+ * @param value The expected write count increment.
+ */
+void sec_mmio_write_increment(uint32_t value);
+
+/**
+ * Checks the expected list of register values.
+ *
+ * All expected register values are verified against expectations. A shutdown
+ * sequence is initiated if any of the comparison fails.
+ *
+ * The `rnd_offset` parameter can be set to a random value to randomize the
+ * order of reads.
+ *
+ * Calling this function will increment the check function counter on a
+ * successful call.
+ *
+ * @param rnd_offset A random value used to generate a random read sequence.
+ */
+void sec_mmio_check_values(uint32_t rnd_offset);
```

Comment:
How do we expect the caller to generate the rnd_offset in the Mask ROM?

Reply:
Added more details to the stringdoc. The caller is expected to use the entropy source or the csrng driver to generate `rnd_offset`.